File: libchunk/chunk_process.c
#include "chunk.h"
#include <stdio.h>
#include <stdlib.h>

#define LENGTH_D 18


void wrapper(char*** chunk, int width, int height, int depth,
    int x, int y, int z,
    char target_block, char shell_block) {
    for (int i = -1; i <= 1; i++) {
        for (int j = -1; j <= 1; j++) {
            for (int k = -1; k <= 1; k++) {
                // Exclude colturile:
                if (i * j * k != 0)
                    continue;
                // Exclude poizitia centrala:
                if (i == 0 && j == 0 && k == 0)
                    continue;
                if (!is_inside(width, height, depth, x + i, y + j, z + k))
                    continue;
                if (chunk[x + i][y + j][z + k] == target_block)
                    continue;
                chunk[x + i][y + j][z + k] = shell_block;
            }
        }
    }
}

char*** chunk_shell(
    char*** chunk, int width, int height, int depth,
    char target_block, char shell_block) {
    int **queue_points = NULL;
    int num_points = 0;
    for (int x = 0; x < width; x++) {
        for (int y = 0; y < height; y++) {
            for (int z = 0; z < depth; z++) {
                if (chunk[x][y][z] != target_block)
                    continue;
                num_points += 1;
                queue_points = realloc(queue_points, num_points * sizeof(int **));
                queue_points[num_points - 1] = malloc(3 * sizeof(int));
                queue_points[num_points - 1][0] = x;
                queue_points[num_points - 1][1] = y;
                queue_points[num_points - 1][2] = z;
            }
        }
    }

    for (int i = 0; i < num_points; i++) {
        int px = queue_points[i][0];
        int py = queue_points[i][1];
        int pz = queue_points[i][2];
        wrapper(chunk, width, height, depth,
            px, py, pz,
            target_block, shell_block);
        free(queue_points[i]);
    }

    free(queue_points);
    return chunk;
}


void fill_algorithm_xOz(char*** chunk, int width, int height, int depth,
    int x, int y, int z, int target_block, int new_block) {
    if (!is_inside(width, height, depth, x, y, z))
        return;
    
    if (chunk[x][y][z] != target_block)
        return;
    chunk[x][y][z] = new_block;
    
    int dx[] = {0, 0, -1, 1};
    int dz[] = {-1, 1, 0, 0};

    for (int i = 0; i < 4; i++) {
        fill_algorithm_xOz(chunk, width, height, depth,
            x+dx[i], y, z+dz[i], target_block, new_block);
    }
}


char*** chunk_fill_xz(
    char*** chunk, int width, int height, int depth,
    int x, int y, int z, char block) {
    if (!is_inside(width, height, depth, x, y, z))
        return chunk;
    
    fill_algorithm_xOz(chunk, width, height, depth, x, y, z, chunk[x][y][z], block);
    return chunk;
}


void fill_algorithm_3D(char*** chunk, int width, int height, int depth,
    int x, int y, int z, int target_block, int new_block) {
    if (!is_inside(width, height, depth, x, y, z))
        return;
    
    if (chunk[x][y][z] != target_block)
        return;
    chunk[x][y][z] = new_block;
    
    int dx[] = {0, 0, 0, 0, -1, 1};
    int dy[] = {0, 0, -1, 1, 0, 0};
    int dz[] = {-1, 1, 0, 0, 0, 0};

    for (int i = 0; i < 6; i++) {
        fill_algorithm_3D(chunk, width, height, depth,
            x+dx[i], y+dy[i], z+dz[i], target_block, new_block);
    }
}

char*** chunk_fill(
    char*** chunk, int width, int height, int depth,
    int x, int y, int z, char block) {
    if (!is_inside(width, height, depth, x, y, z))
        return chunk;

    if (chunk[x][y][z] == block)
        return chunk;
    fill_algorithm_3D(chunk, width, height, depth, x, y, z, chunk[x][y][z], block);
    return chunk;
}

File: method-rs/src/libchunk/chunk_process.rs
use crate::libchunk::chunk_gen::is_inside;

fn wrapper(
    chunk: &mut Vec<Vec<Vec<u8>>>,
    width: usize,
    height: usize,
    depth: usize,
    x: usize,
    y: usize,
    z: usize,
    target_block: u8,
    shell_block: u8,
) {
    for i in -1..=1 {
        let px = (x as isize) + (i as isize);
        if px < 0 {
            continue;
        }

        for j in -1..=1 {
            let py = (y as isize) + (j as isize);
            if py < 0 {
                continue;
            }

            for k in -1..=1 {
                let pz = (z as isize) + (k as isize);
                if pz < 0 {
                    continue;
                }
                if i * j * k != 0 {
                    continue;
                }
                if i == 0 && j == 0 && k == 0 {
                    continue;
                }

                if !is_inside(width, height, depth, px, py, pz) {
                    continue;
                }

                let nx = px as usize;
                let ny = py as usize;
                let nz = pz as usize;

                if chunk[nx][ny][nz] == target_block {
                    continue;
                }
                chunk[nx][ny][nz] = shell_block;
            }
        }
    }
}

pub fn chunk_shell(
    chunk: &mut Vec<Vec<Vec<u8>>>,
    width: usize,
    height: usize,
    depth: usize,
    target_block: u8,
    shell_block: u8,
) {
    let mut points_stack: Vec<Vec<usize>> = vec![];

    for x in 0..width {
        for y in 0..height {
            for z in 0..depth {
                if chunk[x][y][z] != target_block {
                    continue;
                }

                points_stack.push(vec![x, y, z]);
            }
        }
    }

    while !points_stack.is_empty() {
        let pct = points_stack.pop().expect("[ERROR] Empty stack");
        let px = pct[0];
        let py = pct[1];
        let pz = pct[2];

        wrapper(
            chunk,
            width,
            height,
            depth,
            px,
            py,
            pz,
            target_block,
            shell_block,
        );
    }
}

fn fill_algorithm_x0z(
    chunk: &mut Vec<Vec<Vec<u8>>>,
    width: usize,
    height: usize,
    depth: usize,
    x: isize,
    y: isize,
    z: isize,
    target_block: u8,
    new_block: u8,
) {
    if !is_inside(width, height, depth, x as isize, y as isize, z as isize) {
        return;
    }

    if chunk[x as usize][y as usize][z as usize] != target_block {
        return;
    }

    chunk[x as usize][y as usize][z as usize] = new_block;
    let dx: Vec<isize> = vec![0, 0, -1, 1];
    let dz: Vec<isize> = vec![-1, 1, 0, 0];

    for i in 0..dx.len() {
        let nx = (x as isize) + dx[i];
        let nz = (z as isize) + dz[i];

        if nx < 0 || nz < 0 {
            continue;
        }

        fill_algorithm_x0z(
            chunk,
            width,
            height,
            depth,
            nx,
            y,
            nz,
            target_block,
            new_block,
        );
    }
}

pub fn chunk_fill_xz(
    chunk: &mut Vec<Vec<Vec<u8>>>,
    width: usize,
    height: usize,
    depth: usize,
    x: isize,
    y: isize,
    z: isize,
    block: u8,
) {
    if !is_inside(width, height, depth, x, y, z) {
        return;
    }

    fill_algorithm_x0z(
        chunk,
        width,
        height,
        depth,
        x,
        y,
        z,
        chunk[x as usize][y as usize][z as usize],
        block,
    );
}

fn fill_algorithm_3d(
    chunk: &mut Vec<Vec<Vec<u8>>>,
    width: usize,
    height: usize,
    depth: usize,
    x: isize,
    y: isize,
    z: isize,
    target_block: u8,
    new_block: u8,
) {
    if !is_inside(width, height, depth, x, y, z) {
        return;
    }
    if chunk[x as usize][y as usize][z as usize] != target_block {
        return;
    }

    chunk[x as usize][y as usize][z as usize] = new_block;

    let dx: Vec<isize> = vec![0, 0, 0, 0, -1, 1];
    let dy: Vec<isize> = vec![0, 0, -1, 1, 0, 0];
    let dz: Vec<isize> = vec![-1, 1, 0, 0, 0, 0];

    for i in 0..dx.len() {
        let nx = x + dx[i];
        let ny = y + dy[i];
        let nz = z + dz[i];

        if nx < 0 || ny < 0 || nz < 0 {
            continue;
        }

        fill_algorithm_3d(
            chunk,
            width,
            height,
            depth,
            nx,
            ny,
            nz,
            target_block,
            new_block,
        );
    }
}

pub fn chunk_fill(
    chunk: &mut Vec<Vec<Vec<u8>>>,
    width: usize,
    height: usize,
    depth: usize,
    x: isize,
    y: isize,
    z: isize,
    block: u8,
) {
    if !is_inside(width, height, depth, x as isize, y as isize, z as isize) {
        return;
    }
    if chunk[x as usize][y as usize][z as usize] == block {
        return;
    }

    fill_algorithm_3d(
        chunk,
        width,
        height,
        depth,
        x,
        y,
        z,
        chunk[x as usize][y as usize][z as usize],
        block,
    );
}
